Core/Core/Providers/SocialAuth/AppleAuthProvider.swift

```diff
@@ -0,0 +1,103 @@
+//
+//  AppleSingInProvider.swift
+//  Core
+//
+//  Created by Eugene Yatsenko on 10.10.2023.
+//
+
+import Foundation
+import AuthenticationServices
+import Swinject
+
+public struct AppleCredentials: Codable {
+    public var name: String
+    public var email: String
+    public var token: String
+}
+
+public final class AppleAuthProvider: NSObject, ASAuthorizationControllerDelegate {
+
+    private let config: ConfigProtocol
+
+    public init(config: ConfigProtocol) {
+        self.config = config
+        super.init()
+    }
+
+    private var completion: ((Result<AppleCredentials, Error>) -> Void)?
+    private let appleIDProvider = ASAuthorizationAppleIDProvider()
+
+    public func request(completion: ((Result<AppleCredentials, Error>) -> Void)?) {
+        let request = appleIDProvider.createRequest()
+        request.requestedScopes = [.fullName, .email]
+
+        let authorizationController = ASAuthorizationController(authorizationRequests: [request])
+        authorizationController.delegate = self
+        authorizationController.performRequests()
+
+        self.completion = completion
+    }
+
+    public func authorizationController(
+        controller: ASAuthorizationController,
+        didCompleteWithAuthorization authorization: ASAuthorization
+    ) {
+        guard let credentials = authorization.credential as? ASAuthorizationAppleIDCredential else {
+            completion?(.failure(SocialAuthError.unknownError))
+            return
+        }
+
+        var storage = Container.shared.resolve(CoreStorage.self)
+        let pncf = PersonNameComponentsFormatter()
+
+        var name = storage?.appleSignFullName ?? ""
+        if let fullName = credentials.fullName {
+            name = pncf.string(from: fullName)
+        }
+
+        let email = credentials.email ?? storage?.appleSignEmail ?? ""
+
+        if !name.isEmpty {
+            storage?.appleSignFullName = name
+        }
+
+        if storage?.appleSignEmail == nil {
```

Comment:
The value of `storage?.appleSignEmail` is an empty string so the email is not being saved into storage for later ussage.